csv_permissions/permissions.py

```diff
@@ -182,168 +64,253 @@ def _parse_csv(
+    permission_to_user_type_to_evaluator: Dict[PermName, Dict[UserType, Evaluator]] = {}
+    for permission, user_type_to_unresolved in permission_to_user_type_to_unresolved.items():
+        if permission not in permission_to_user_type_to_evaluator:
+            permission_to_user_type_to_evaluator[permission] = {}
+        for user_type, detail in user_type_to_unresolved.items():
+            try:
+                for resolve_evaluator in resolve_evaluators:
+                    evaluator = resolve_evaluator(detail)
+                    if evaluator is not None:
+                        permission_to_user_type_to_evaluator[permission][user_type] = evaluator
+                        break
+                else:
+                    raise ValueError(f"Could not resolve {permission} for {user_type} to anything")
+            except Exception as e:
+                raise RuntimeError(f"Error resolving {permission} for {user_type}: {detail.evaluator_name} ({e})") from e
 
-                function_name = custom
-                access_level = "custom"
+    return permission_to_user_type_to_evaluator, permission_is_global, known_user_types, known_perms
 
-            elif access_level in ("all", "", "yes"):
-                pass
 
-            else:
-                warnings.warn(f"{access_level} is not a valid access level.")
-                continue
+# note that django creates a new instance of an auth backend for every permission check!
+class CSVPermissionsBackend:
+    permission_lookup: Dict[PermName, Dict[UserType, Evaluator]]
+    permission_is_global: Dict[PermName, bool]
+    known_user_types: Set[UserType]
+    known_perms: Set[PermName]
 
+    def __init__(self):
+        try:
+            permissions_paths = settings.CSV_PERMISSIONS_PATHS
+        except AttributeError:
             try:
-                permission_to_group_to_function[permission][group] = _resolve_function(
-                    detail.app_config, access_level, function_name, is_global
+                settings.CSV_PERMISSIONS_PATHS = (settings.CSV_PERMISSIONS_PATH,)
+            except AttributeError:
+                raise ImproperlyConfigured("csv_permissions requires settings.CSV_PERMISSIONS_PATHS to be set")
+            else:
+                warnings.warn(
+                    "settings.CSV_PERMISSIONS_PATH is deprecated in favor of settings.CSV_PERMISSIONS_PATHS",
+                    DeprecationWarning
                 )
-            except RuntimeError as e:
-                raise RuntimeError(f"{e} Permission: {permission}")
+                permissions_paths = settings.CSV_PERMISSIONS_PATHS
 
-    return permission_to_group_to_function, permission_is_global
+        # make sure it's always a tuple so that it's hashable and _resolve_functions() can have @lru_cache() applied
+        if not isinstance(permissions_paths, tuple):
+            permissions_paths = tuple(permissions_paths)
+            settings.CSV_PERMISSIONS_PATHS = permissions_paths
```

Comment:
i think maybe either support it to be a not iterable and turn it into a tuple, or throw when its not iterable - mainly to catch people doing settings.CSV_PERMISSIONS_PATHS="x.csv"

Reply:
(Leaving this as a placeholder comment, please don't resolve this comment yet)

Every single time you do a permissions check django instantiates a new copy of the permissions backend. When the backend does nothing this is fine, but there's a lot csv permissions needs to do to load the files. I've worked around this by putting an `@lru_cache` on `_resolve_functions` and having the constructor for `CsvPermissionsBackend` do as little as possible. If we allow both settings then we either need to do extra work on each CsvPermissionsBackend init, or we need to have some way to ensure that a cached CSV_PERMISSIONS_PATHS doesn't override any updates to CSV_PERMISSIONS_PATH in the tests.

TODO: if we use `@override_settings` it should revert any changes to django `settings`, but does restore individual settings or the whole settings object?

Reply:
Have changed it so that `CSV_PERMISSION_PATHS` (list/tuple of paths) and `CSV_PERMISSION_PATH` (single path) are now both allowed